tests/test_pyvsi.py

```diff
@@ -0,0 +1,176 @@
+    with rasterio.open(rgb_file_object) as src:
+        assert src.driver == 'GTiff'
+        assert src.count == 3
+        assert src.dtypes == ('uint8', 'uint8', 'uint8')
+        assert src.read().shape == (3, 718, 791)
+
+
+def test_file_object_read_variant(rgb_file_object):
+    """An example of reading from a PythonVSIFile object"""
+    with rasterio.open(PythonVSIFile(rgb_file_object)) as src:
+        assert src.driver == 'GTiff'
+        assert src.count == 3
+        assert src.dtypes == ('uint8', 'uint8', 'uint8')
+        assert src.read().shape == (3, 718, 791)
+
+
+def test_file_object_read_variant2(rgb_file_object):
+    """An example of reading from a BytesIO object version of a file's contents."""
+    with rasterio.open(BytesIO(rgb_file_object.read())) as src:
+        assert src.driver == 'GTiff'
+        assert src.count == 3
+        assert src.dtypes == ('uint8', 'uint8', 'uint8')
+        assert src.read().shape == (3, 718, 791)
+
+
+def test_vrt_vsifile(data_dir, path_white_gemini_iv_vrt):
+    """Successfully read an in-memory VRT"""
+    with open(path_white_gemini_iv_vrt) as vrtfile:
+        source = vrtfile.read()
+        source = source.replace('<SourceFilename relativeToVRT="1">389225main_sw_1965_1024.jpg</SourceFilename>', '<SourceFilename relativeToVRT="0">{}/389225main_sw_1965_1024.jpg</SourceFilename>'.format(data_dir))
+        source = BytesIO(source.encode('utf-8'))
+
+    with PythonVSIFile(source) as vsifile:
+        with vsifile.open() as src:
+            assert src.driver == 'VRT'
+            assert src.count == 3
+            assert src.dtypes == ('uint8', 'uint8', 'uint8')
+            assert src.read().shape == (3, 768, 1024)
+
+
+@pytest.mark.xfail(reason="Copying is not supported by PythonVSIFile")
+def test_vsifile_copyfiles(path_rgb_msk_byte_tif):
+    """Multiple files can be copied to a PythonVSIFile using copyfiles"""
+    with rasterio.open(path_rgb_msk_byte_tif) as src:
+        src_basename = os.path.basename(src.name)
+        with PythonVSIFile(dirname="foo", filename=src_basename) as vsifile:
+            copyfiles(src.name, vsifile.name)
+            with vsifile.open() as rgb2:
+                assert sorted(rgb2.files) == sorted(['/vsimem/foo/{}'.format(src_basename), '/vsimem/foo/{}.msk'.format(src_basename)])
+
+
+@pytest.mark.xfail(reason="PythonVSIFile does not implement '.files' property properly.")
+def test_multi_vsifile(path_rgb_msk_byte_tif):
+    """Multiple files can be copied to a PythonVSIFile using copyfiles"""
+    with open(path_rgb_msk_byte_tif, 'rb') as tif_fp, open(path_rgb_msk_byte_tif + '.msk', 'rb') as msk_fp:
+        with PythonVSIFile(tif_fp, dirname="bar", filename='foo.tif') as tifvsifile, \
+                PythonVSIFile(msk_fp, dirname="bar", filename='foo.tif.msk') as mskvsifile:
+            with tifvsifile.open() as src:
+                assert sorted(os.path.basename(fn) for fn in src.files) == sorted(['foo.tif', 'foo.tif.msk'])
```

Comment:
@sgillies Where is this `.files` property defined? I can't seem to find it anywhere (this was copied from the memory file tests).